datastore/gcloud/aio/datastore/filter.py

```diff
@@ -0,0 +1,111 @@
+from typing import Any
+from typing import Dict
+from typing import List
+
+from gcloud.aio.datastore.constants import CompositeFilterOperator
+from gcloud.aio.datastore.constants import PropertyFilterOperator
+from gcloud.aio.datastore.value import Value
+
+
+class _BaseFilter:
+    json_key = ''
+
+    def __repr__(self) -> str:
+        return str(self.to_repr())
+
+    @classmethod
+    def from_repr(cls, data: Dict[str, Any]) -> '_BaseFilter':
+        raise NotImplementedError
+
+    def to_repr(self) -> Dict[str, Any]:
+        raise NotImplementedError
+
+
+# https://cloud.google.com/datastore/docs/reference/data/rest/v1/projects/runQuery#Filter
+class Filter:
+    def __init__(self, inner_filter: _BaseFilter):
+        self.inner_filter = inner_filter
+
+    def __repr__(self) -> str:
+        return str(self.to_repr())
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Filter):
+            return False
+        return self.inner_filter == other.inner_filter
+
+    @classmethod
+    def from_repr(cls, data: Dict[str, Any]) -> 'Filter':
+        (inner_filter_name, inner_filter_data), = data.items()
```

Comment:
We generally prefer using early returns a bit more to try to keep the flows clear. I think using key lookups directly rather than doing the clever deserialization would be a bit more readable:

```
    @classmethod
    def from_repr(cls, data: Dict[str, Any]) -> 'Filter':
        if 'compositeFilter' in data:
            return cls(CompositeFilter.from_repr(data['compositeFilter']))
        if 'propertyFilter' in data:
            return cls(PropertyFilter.from_repr(data['propertyFilter']))

        raise ValueError(f'invalid filter name: {data.keys()}')
```

Reply:
Ah yes, that's much more readable. I think I was trying too hard to not repeat myself